src/Symfony/Component/Validator/ConstraintValidator.php

```diff
@@ -44,6 +45,40 @@ public function initialize(ExecutionContextInterface $context)
         $this->context = $context;
     }
 
+    /**
+     * Test a constraint class for the current validator.
+     *
+     * @throws UnexpectedTypeException
+     */
+    final protected static function testConstraint(Constraint $constraint, string $expectedClass)
```

Comment:
do we really need this? I feel like this removed some info for static analysis + increases the stack traces of exception for little practical benefit, don't you think?

Reply:
I ran a little investigation:

1. PHPStorm doesn't understand this
2. Psalm does https://getpsalm.org/r/acd72ab0ab cc @muglug  
3. PHPStan doesn't https://phpstan.org/r/be6b68a728fc1d24fb27069640f46701 cc @ondrejmirtes 

Anyway Validator is full of this piece of code, so I think it's worth it. SA tools which don't support it is problem of the tool itself, not the code. And we can help them by specifying `@param` alongside `@inheritDoc` (works for PHPStorm and Psalm, PHPStan is stubborn like usual).

Extraction of common pieces of code is important for keeping consistency. Also, if this was done in the first place, there would be no need to manually touch so many places in #27917

But, this should be renamed. `testConstraint` is odd one. `assertInstanceOf`?

Reply:
That's not really a common piece of logic - that's an "if"...

Reply:
I didn't mention "logic", I mentioned "piece of code". Anyway "if" contains some logic. This "if" is in almost all validators. Devs keep producing highly repetitive code instead of extracting it to function. This is like no. 1 rule of programming. Programs are good at repeating, not humans.

And here is another advantage of having such thing there in the first place - if it was already there, all custom validators using such function would be automatically compatible with #27917. Since it's not there (because such function is too simple right?), every custom constraint will now have to change this piece of code within it. But it's too late for #27917. Maybe another time there will be need to change the message or exception type.

Reply:
I get that, but I don't agree: designing the code to make such changes easy is not a target at all to me. Software architecture should be built around things that change, not those that don't (#27917 is not a change that should be centralized to me.)